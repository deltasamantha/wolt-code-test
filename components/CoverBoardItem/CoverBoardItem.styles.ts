import {StyleSheet} from "react-native";

export const styles = StyleSheet.create({
  item: {
    position: "absolute",
  },
});
