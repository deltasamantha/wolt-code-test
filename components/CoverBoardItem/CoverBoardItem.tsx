import React, {useEffect} from "react";
import {Image, ImageSourcePropType, ImageStyle, StyleProp} from "react-native";
import Animated, {
  Extrapolate,
  interpolate,
  useAnimatedStyle,
  useDerivedValue,
  useSharedValue,
  withTiming,
} from "react-native-reanimated";

import {styles} from "./CoverBoardItem.styles";
import {timingConfig} from "../../config/AnimationConfig";

interface Props {
  originX: number;
  originY: number;
  destinationX: number;
  destinationY: number;
  source: ImageSourcePropType;
  style: StyleProp<ImageStyle>;
  testID?: string;
}

const AnimatedImage = Animated.createAnimatedComponent(Image);

const CoverBoardItem: React.FC<Props> = ({
  originX,
  originY,
  destinationX,
  destinationY,
  source,
  style,
  testID,
}: Props) => {
  const translation = useSharedValue<number>(0);

  useEffect(() => {
    translation.value = 1;
  }, []);

  const plate1TranslateX = useDerivedValue<number>(() => {
    return withTiming(
      interpolate(translation.value, [0, 1], [originX, 0], Extrapolate.CLAMP),
      timingConfig,
    );
  });
  const plate1TranslateY = useDerivedValue<number>(() => {
    return withTiming(
      interpolate(translation.value, [0, 1], [originY, 0], Extrapolate.CLAMP),
      timingConfig,
    );
  });

  const plate1Animation = useAnimatedStyle(() => {
    return {
      transform: [
        {translateX: plate1TranslateX.value},
        {translateY: plate1TranslateY.value},
      ],
    };
  });
  return (
    <AnimatedImage
      source={source}
      testID={`${testID}`}
      style={[
        styles.item,
        style,
        {left: destinationX, top: destinationY},
        plate1Animation,
      ]}
    />
  );
};

export default CoverBoardItem;
