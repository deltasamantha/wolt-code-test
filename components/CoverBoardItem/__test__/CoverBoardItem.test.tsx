import {render, cleanup, screen} from "@testing-library/react-native";

import CoverBoardItem from "../CoverBoardItem";

afterEach(cleanup);

describe("CoverBoardItem", () => {
  it("Should render CoverBoardItem properly", () => {
    const {toJSON} = render(
      <CoverBoardItem
        source={require("../../../assets/images/food-plate-top.png")}
        originX={0}
        originY={0}
        destinationX={10}
        destinationY={10}
        style={{width: 100, height: 100}}
      />,
    );
    expect(toJSON()).toMatchSnapshot();
  });

  it("Should contains an Image component", () => {
    render(
      <CoverBoardItem
        source={require("../../../assets/images/food-plate-top.png")}
        originX={0}
        originY={0}
        destinationX={10}
        destinationY={10}
        style={{width: 100, height: 100}}
      />,
    );
    const imageItem = screen.getByTestId("board-item", {exact: false});
    expect(imageItem).toBeTruthy();
  });
});
