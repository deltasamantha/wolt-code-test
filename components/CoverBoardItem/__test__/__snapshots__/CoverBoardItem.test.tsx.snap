// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CoverBoardItem Should render CoverBoardItem properly 1`] = `
<Image
  source={1}
  style={
    [
      {
        "position": "absolute",
      },
      {
        "height": 100,
        "width": 100,
      },
      {
        "left": 10,
        "top": 10,
      },
      {
        "transform": [
          {
            "translateX": undefined,
          },
          {
            "translateY": undefined,
          },
        ],
      },
    ]
  }
  testID="undefined-board-item"
/>
`;
